Issue #6207: Track remaing checks to do in xpath IT

mincong-h opened the pull request:
This unit test aims to ~~provide validation~~ keep track of remaining check on XPath related integration itests (IT). This idea is proposed by @romani in https://github.com/checkstyle/checkstyle/issues/6207#issuecomment-521500309 :

> it would be awesome is you create UT with list like - https://github.com/checkstyle/checkstyle/pull/6969/files#diff-67a0806621fe751520ab842060ba14fcL106 , see whole change of #6969 it show how @pbludov improve step by step , but it is very enforced validation on each change, new Checks already come in better form in first commit.
so list of what is not done will be self controlled by UTs and nobody will forget to add such tests during new Check is created. We recently missed such tests creation ... .

rnveach commented:
@mincong-h CI is failing.  Please fix.

> Validate package names in xpath IT 

This wording doesn't sound right. You aren't validating the `package` statement in the Java code or checking the naming of the folders. Like you said above you are just `keeps track of all the missing checks to be tested` in the regression area.

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/XpathRegressionTest.java:
```diff
@@ -0,0 +1,236 @@
+
+public class XpathRegressionTest extends AbstractModuleTestSupport {
+
+    private static final Set<String> CHECK_NAMES = new HashSet<>(Arrays.asList(
```
I don't see why we have a list of hardcoded check names. Is this just a list of all the checks by checkstyle in our repo?
If so, this should be using our util.
https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/CheckUtil.java#L138

mincong-h replied:
> I don't see why we have a list of hardcoded check names. Is this just a list of all the checks by checkstyle in our repo?

Yes, right. I just tried to have a complete list of checks, so that I can keep track of the remaining tasks. I went through class PackageObjectFactory, but that map is private:

https://github.com/checkstyle/checkstyle/blob/c5cc85e47ddd63c85a56c32ae26f0ad6f5310874/src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java#L99

> https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/CheckUtil.java#L138

Thanks for your suggestion!

mincong-h commented:
> @mincong-h CI is failing. Please fix.

Yes, right. Let me do that.
 
> > Validate package names in xpath IT
> 
> This wording doesn't sound right. You aren't validating the `package` statement in the Java code or checking the naming of the folders. Like you said above you are just `keeps track of all the missing checks to be tested` in the regression area.

Yes, correct. I'm not validating the `package` statements in Java code at all. However, I do check the folder naming, by listing the folders located in "src/it/resources/org/checkstyle/suppressionxpathfilter". Is it OK as a first step?

rnveach commented:
This is all great, I just have an small issue with the wording is all.

> I do check the folder naming, by listing the folders located in "src/it/resources/org/checkstyle/suppressionxpathfilter".

Don't you just list the files in the folders? Besides making sure that folder exists with files in it, you aren't do any real validation of the folder naming which is the heart of the package name. Please correct me if I am wrong.
I just recommend naming this like `Issue #6207: Validate all checks have tests in xpath IT`

mincong-h commented:
> Don't you just list the files in the folders? Besides making sure that folder exists with files in it, you aren't do any real validation of the folder naming which is the heart of the package name. Please correct me if I am wrong.

Not really. I list all the directories of the "suppressionxpathfilter", which is similar to:

```
$ find ./src/it/resources/org/checkstyle/suppressionxpathfilter -depth 1
```

Then I make sure that each directory has name matching a Checkstyle check. Creating any directory which does not match it will fail the build. I'm not listing regular files (Java files) at all. I think the most confusing part comes from the misuse of word "package" in my commit message, my code, and in this PR...

mincong-h commented:
I updated my commit, I hope it makes things clearer

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/XpathUtil.java:
```diff
@@ -36,6 +36,9 @@
 public final class XpathUtil {
 
+    public static final String IT_RESOURCES_XPATH_FILTER_PATH =
+            "src/it/resources/org/checkstyle/suppressionxpathfilter";
```
Can't the test class extend something like https://github.com/checkstyle/checkstyle/blob/master/src/it/java/org/checkstyle/base/AbstractItModuleTestSupport.java and then call `getPath("")`?
https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java#L57

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/XpathRegressionTest.java:
```diff
@@ -0,0 +1,140 @@
+                if (Files.isDirectory(path)) {
+                    final String dirName = path.toFile().getName();
+                    if (!allowedDirectoryNames.contains(dirName)) {
+                        fail("Invalid directory name: " + dirName);
```
Why not `assertTrue` and remove the condition/fail?

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/XpathRegressionTest.java:
```diff
@@ -0,0 +1,140 @@
+                        fail("Invalid directory name: " + dirName);
+                    }
+                    if (missingDirectoryNames.contains(dirName)) {
+                        fail("Check is now tested as '" + dirName + "', please update"
```
same.

mincong-h commented:
> I see where my confusion is coming from now. `MISSING_CHECK_NAMES` made me think we were checking for missing tests, but you are really checking for missing inputs for a test.
> Usually we confirm the test exists, not the input. Doesn't it make more sense that we do that?

Yes, agree with you. Actually, I didn't check the existing tests.

> Something like confirming the resources are in the correct location was for something like `AllTestsTest`.
> See
> 
> https://github.com/checkstyle/checkstyle/blob/a73ff0890c27cbe8affc1a55f2710a231687bd85/src/test/java/com/puppycrawl/tools/checkstyle/internal/AllTestsTest.java#L54

Before submitting another commit, just want to discuss the goal to be sure we are aligned. What we need to do here is:

- Keep track of the remaining tasks as a unit test
- Check existence of tests (Java files) match the naming convention:
    ```
    XpathRegression${checkName}Test
    ```

However, checking input files (resources) is not needed.

rnveach commented:
@romani Please confirm too before we do any more.

I myself agree with all your points, except:

> However, checking input files (resources) is not needed. Is that correct?

IMO, we should verify inputs are connected to a test, named correctly, and in the right folder. This doesn't have to be done in this PR.
My recommendation is we already have a test setup for doing this, but it is built only for Test area and not IT. That is what https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/internal/AllTestsTest.java#L54 is . I think the IT area should be added to that existing test.
I started a branch doing this at https://github.com/rnveach/checkstyle/commits/it_input_rename but it is not complete. If you want to complete it, then be my guest.

romani commented:
> However, checking input files (resources) is not needed. Is that correct

ideally, it would be awesome to have a format for inputs too - `SuppressionXpathRegression{Check name}.*\.java` . But it is completely fine to deliver it as separate PR, or could be done a bit later it if is convenient.

mincong-h commented:
> @romani Please confirm too before we do any more.
> 
> I myself agree with all your points, except:
> 
> > However, checking input files (resources) is not needed. Is that correct?
> 
> IMO, we should verify inputs are connected to a test, named correctly, and in the right folder. This doesn't have to be done in this PR.

I updated my test to include those.

> My recommendation is we already have a test setup for doing this, but it is built only for Test area and not IT. That is what https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/internal/AllTestsTest.java#L54 is . I think the IT area should be added to that existing test.

Sorry I missed this part... I didn't include the logic in `AllTestsTest`. Will do later this weekend.

> I started a branch doing this at https://github.com/rnveach/checkstyle/commits/it_input_rename but it is not complete. If you want to complete it, then be my guest.

Will check, thanks!

>  > However, checking input files (resources) is not needed. Is that correct
>
> ideally, it would be awesome to have a format for inputs too - `SuppressionXpathRegression{Check name}.*\.java` .

Good idea! I updated my PR (will squash commits afterwards)

romani commented:
@mincong-h , it is better to reply "done" to our comments, in this case we get notification that you are done and ready for review.  It is better to avoid "I Will ..." , it is better to disagree or simply reply as "done" when you are ready.
Such simple rules will speedup PR review.

romani commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/XpathRegressionTest.java:
```diff
@@ -0,0 +1,208 @@
+public class XpathRegressionTest extends AbstractModuleTestSupport {
+
+    // Expand XPath IT Regression Testing
+    // https://github.com/checkstyle/checkstyle/issues/6207
```
it is better to have comment `Temporal list of Checks that allowed to have no XPath IT Regression Testing`

mincong-h replied:
Done. I just removed word "list of" from your suggestion, because literally this is a set 😅 . So, the finally comment is: _"Temporal Checks that allowed to have no XPath IT Regression Testing"_

romani commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/XpathRegressionTest.java:
```diff
@@ -0,0 +1,208 @@
+                assertTrue("Unknown check '" + check + "' in test file: " + filename,
+                        simpleCheckNames.contains(check));
+
+                assertFalse("Check '" + check + "' is now tested. Please update the todo list.",
```
what is todo list ?
it is better to reference as `{Classname}.{field}`

mincong-h replied:
done

romani commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/XpathRegressionTest.java:
```diff
@@ -0,0 +1,208 @@
+    }
+
+    @Test
+    public void validateJavaFiles() throws Exception {
```
`validateIntegrationTestClassNames` I think would be better.

mincong-h replied:
done

mincong-h commented:
> @mincong-h , it is better to reply "done" to our comments, in this case we get notification that you are done and ready for review. It is better to avoid "I Will ..." , it is better to disagree or simply reply as "done" when you are ready.
> Such simple rules will speedup PR review.

+1. Thanks for your suggestions!

romani commented:
Please squash all commits at one, be ready for merge.

mincong-h commented:
> IMO, we should verify inputs are connected to a test, named correctly, and in the right folder. This doesn't have to be done in this PR.
My recommendation is we already have a test setup for doing this, but it is built only for Test area and not IT. That is what https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/internal/AllTestsTest.java#L54 is . I think the IT area should be added to that existing test.

@rnveach I just checked `AllTestsTest.java`, it's a good idea. But as you mentioned, I would like to it done in a separated PR. Having it here will mean:

1. Make the review more difficult
2. Checking all ITs input files by walking through the IT resource directory
3. Rename the XPath regressions files, so that they start with `Input`

For now, I use a simplified version here `XpathRegressionTest#validateInputFiles()` to check input files.

> I started a branch doing this at https://github.com/rnveach/checkstyle/commits/it_input_rename but it is not complete. If you want to complete it, then be my guest.

I would like to finish #6207 first, then if time allowed, contribute to the renaming one.

mincong-h commented:
> Please squash all commits at one, be ready for merge.

Done.

rnveach commented:
I am fine with it not being in this PR. I always meant for it to be for a future PR.

If you plan to work on that branch, I do however question why this PR needs to do similar functionality.
If the other test will validate the input files and you will finish it, then why do we need the `validateInputFiles` test in this PR? Isn't it basically going to amount to duplicating the same work?

mincong-h commented:
> I am fine with it not being in this PR. I always meant for it to be for a future PR.
> 
> If you plan to work on that branch, I do however question why this PR needs to do similar functionality.
> If the other test will validate the input files and you will finish it, then why do we need the `validateInputFiles` test in this PR? Isn't it basically going to amount to duplicating the same work?

Because I wanted to have a minimum validation for existing XPath ITs before adding more. It helped to correct several input files. I didn't clearly know the role of different tests at the beginning. If you think it's better to split the commits, or drop the validation part, I'm completely fine. I can modify the code.

romani commented:
Let's improve validation step by step.
I recommend to merge this PR sooner
